internal/gatewayapi/route.go

```diff
@@ -209,7 +209,7 @@ func (t *Translator) processHTTPRouteRules(httpRoute *HTTPRouteContext, parentRe
 
 		// If the route has no valid backends then just use a direct response and don't fuss with weighted responses
 		for _, ruleRoute := range ruleRoutes {
-			if ruleRoute.BackendWeights.Invalid > 0 && ruleRoute.Destination == nil {
+			if (ruleRoute.BackendWeights.Invalid > 0 || len(rule.BackendRefs) == 0) && ruleRoute.Destination == nil {
```

Comment:
why isnt it hitting `ruleRoute.Destination == nil` ? that feels like the bug

Reply:
should it be a `or` / `||` instead ?

Reply:
`ruleRoute.BackendWeights.Invalid` is only incremented when `BackendRefs` is non-empty as a result of being set during destination processing, so it would seem to be an intentional gate. In the case of an empty `BackendRefs` though, `ruleRoute.BackendWeights.Invalid` will always be zero which is the cause of failure when publishing a route with no backend refs nor filters.

If I remove `ruleRoute.BackendWeights.Invalid > 0 &&` or change it to `||` instead, a handful of existing tests have `directResponse` show up in `xdsIR` output. For example, this one: https://github.com/envoyproxy/gateway/blob/main/internal/gatewayapi/testdata/httproute-with-redirect-filter-hostname.in.yaml although maybe that is a poorly written test since applying that object would normally result in a `RequestRedirect filter must not be used together with backendRefs` validation error 🤔

Reply:
ah, cool, lets stick with the logic in this PR then

Reply:
There is actually a problem with my proposed logic, since `backendRefs` cannot be used with the `RequestRedirect` filter it would break this route:

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: redirected
spec:
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: eg
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /redirected
    filters:
    - type: RequestRedirect
      requestRedirect:
        scheme: https
        statusCode: 301
        hostname: "redirected.com"
```

Reply:
yeah, all the HTTPRouteRedirect are failing in the conformance tests kinda proving this fix is incomplete: https://github.com/envoyproxy/gateway/actions/runs/8103755213/job/22149232447?pr=2733

I think changing the logic to `if ruleRoute.Destination == nil && ruleRoute.Redirect == nil {` will cover all the cases (and it passed conformance locally) so I'm going to push that up in a minute